fpmsyncd/routesync.cpp

```diff
@@ -886,6 +1004,105 @@ bool RouteSync::getIfName(int if_index, char *if_name, size_t name_len)
     return true;
 }
 
+/*
+ * getNextHopList() - parses next hop list attached to route_obj
+ * @arg route_obj     (input) Netlink route object
+ * @arg gw_list       (output) comma-separated list of NH IP gateways
+ * @arg mpls_list     (output) comma-separated list of NH MPLS info
+ * @arg intf_list     (output) comma-separated list of NH interfaces
+ *
+ * Return bool true=Success false=Failure
+ */
+bool RouteSync::getNextHopList(struct rtnl_route *route_obj, string& gw_list,
+                               string& mpls_list, string& intf_list)
+{
+    bool mpls_found = false;
+
+    for (int i = 0; i < rtnl_route_get_nnexthops(route_obj); i++)
+    {
+        struct rtnl_nexthop *nexthop = rtnl_route_nexthop_n(route_obj, i);
+        struct nl_addr *addr = NULL;
+
+        /* RTA_GATEWAY is NH gateway info for IP routes only */
+        if ((addr = rtnl_route_nh_get_gateway(nexthop)))
+        {
+            char gw_ip[MAX_ADDR_SIZE + 1] = {0};
+            nl_addr2str(addr, gw_ip, MAX_ADDR_SIZE);
+            gw_list += gw_ip;
+        }
+        /* RTA_VIA is NH gateway info for MPLS routes only */
+        else if ((addr = rtnl_route_nh_get_via(nexthop)))
+        {
+            char via_ip[MAX_ADDR_SIZE + 1] = {0};
+            nl_addr2str(addr, via_ip, MAX_ADDR_SIZE);
+            gw_list += via_ip;
+        }
+        else
+        {
+            if (rtnl_route_get_family(route_obj) == AF_INET6)
+            {
+                gw_list += "::";
+            }
+            /* for MPLS route, use IPv4 as default gateway. */
+            else
+            {
+                gw_list += "0.0.0.0";
+            }
+        }
+
+        /* RTA_NEWDST is MPLS NH label stack for MPLS routes only */
+        if ((addr = rtnl_route_nh_get_newdst(nexthop)))
+        {
+            char label[MAX_ADDR_SIZE + 1] = {0};
+            nl_addr2str(addr, label, MAX_ADDR_SIZE);
+            mpls_list += string("push+");
```

Comment:
If I remember correctly, RTA_NEWDST indicates a 'SWAP' operation. Incoming label route is swapped with the label stack in the nexthop.

RTA_ENCAP indicates a push. In this case it can be "label" route or an "IP" route.

Reply:
I have already stated in the review meeting that the fpmsyncd would continue to follow the Linux/FRR model where SWAP is programmed as POP+PUSH.

Reply:
Also, you are incorrect that RTA_ENCAP can be used with label route.  RTA_ENCAP_MPLS can only be used with IP routes.

Reply:
Netlink protocol only provides attributes and leaves the interpretation to the user. Yes, I don't see RTA_ENCAP being used for label routes either by iproute2 or by FRR. It would be interesting to see how linux programs if RTA_ENCAP is sent for label routes. However there is a difference in next hop attributes for IP route and Label route.

https://github.com/FRRouting/frr/blob/525ec2e0734a0d492401a5116b99c5b1af47c34a/zebra/rt_netlink.c#L1150

For IP routes the netlink attributes are: RTA_ENCAP_TYPE, RTA_ENCAP and MPLS_IPTUNNEL_DST

for Label routes it is: RTA_NEWDST

This can also be verified by capturing netlink messages while programming label and IP routes with label nexthops.

Reply:
@smaheshm I said at the review meeting last week that I had already tested this behavior in Linux.  Attempting to program ENCAP_MPLS on AF_MPLS route will fail with error "Error:  Unknown attribute."  This error comes straight from the Linux kernel, it is not an iproute2 error.  Linux kernel does not support what you are proposing.